core/src/main/java/brooklyn/entity/rebind/RebindIteration.java

```diff
@@ -899,48 +914,64 @@ protected void setCatalogItemId(BrooklynObject item, String catalogItemId) {
+                        // This is a dangling reference to the catalog item (which will have been logged by lookupCatalogItem).
+                        // Try loading as any version.
+                        if (CatalogUtils.looksLikeVersionedId(catalogItemId)) {
+                            String symbolicName = CatalogUtils.getIdFromVersionedId(catalogItemId);
+                            catalogItem = rebindContext.lookup().lookupCatalogItem(symbolicName);
+                            
+                            if (catalogItem != null) {
+                                LOG.warn("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId
+                                        +" ("+bType.getSimpleName()+"); will auto-upgrade to "+catalogItem.getCatalogItemId());
+                                catalogItemId = catalogItem.getCatalogItemId();
+                            }
                         }
                     }
-                    throw new IllegalStateException("No catalogItemId specified and can't load class from either classpath of catalog items");
+                }
+                if (catalogItem != null) {
+                    BrooklynClassLoadingContext loader = CatalogUtils.newClassLoadingContext(managementContext, catalogItem);
+                    return new LoadedClass<T>(loader.loadClass(jType, bType), catalogItemId);
                 } else {
-                    throw new IllegalStateException("No catalogItemId specified and can't load class from classpath");
+                    LOG.warn("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId
+                            +" ("+bType.getSimpleName()+"); will try default class loader");
                 }
-
             }
-        }
+            
+            try {
+                return new LoadedClass<T>((Class<T>)reflections.loadClass(jType), catalogItemId);
+            } catch (Exception e) {
+                Exceptions.propagateIfFatal(e);
+                LOG.warn("Unable to load "+jType+" using reflections; will try standard context");
+            }
 
-        protected BrooklynClassLoadingContext getLoadingContextFromCatalogItemId(String catalogItemId, ClassLoader classLoader, RebindContext rebindContext) {
-            Preconditions.checkNotNull(catalogItemId, "catalogItemId required (should not be null)");
-            CatalogItem<?, ?> catalogItem = rebindContext.lookup().lookupCatalogItem(catalogItemId);
-            if (catalogItem != null) {
-                return CatalogUtils.newClassLoadingContext(managementContext, catalogItem);
+            if (catalogItemId != null) {
+                throw new IllegalStateException("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId+", or load class from classpath");
+            } else if (BrooklynFeatureEnablement.isEnabled(BrooklynFeatureEnablement.FEATURE_BACKWARDS_COMPATIBILITY_INFER_CATALOG_ITEM_ON_REBIND)) {
```

Comment:
Looking at the code in details now it's not obvious why this (old) code repeats the logic in `findCatalogItemId`. Actually the whole catalog item inference could be encapsulated there. Just thinking out loud, no need to change it in this PR.